src/machine/machine_rp2040_pwm.go

```diff
@@ -0,0 +1,494 @@
+// measure the frequency or duty cycle of an input signal. This gives a total of up to 16 controllable
+// PWM outputs. All 30 GPIOs can be driven by the PWM block
+//
+// The PWM hardware functions by continuously comparing the input value to a free-running counter. This produces a
+// toggling output where the amount of time spent at the high output level is proportional to the input value. The fraction of
+// time spent at the high signal level is known as the duty cycle of the signal.
+//
+// The default behaviour of a PWM slice is to count upward until the wrap value (\ref pwm_config_set_wrap) is reached, and then
+// immediately wrap to 0. PWM slices also offer a phase-correct mode, where the counter starts to count downward after
+// reaching TOP, until it reaches 0 again.
+type pwms struct {
+	slice pwmGroup
+	hw    *rp.PWM_Type
+}
+
+// Handle to all pwm peripheral registers.
+var _PWM = pwms{
+	hw: rp.PWM,
+}
+
+// Initialise a PWM with settings from a configuration object.
+// If start is true then PWM starts on initialization.
+//go:inline
+func (pwm *pwms) init(sliceNum uint8, config PWMConfig, start bool) {
+	if pwm.slice == nil {
+		pwm.initSlice()
+	}
+
+	// Not enable Phase correction
+	pwm.setPhaseCorrect(sliceNum, false)
+
+	// Clock mode set by default to Free running
+	pwm.setDivMode(sliceNum, rp.PWM_CH0_CSR_DIVMODE_DIV)
+
+	// Set Output polarity (false/false)
+	pwm.setInverting(sliceNum, 0, false)
+	pwm.setInverting(sliceNum, 1, false)
+
+	// Set wrap. The highest value the counter will reach before returning to zero, also known as TOP.
+	pwm.setWrap(sliceNum, 0xffff)
+	// period is set after TOP (Wrap).
+	pwm.setPeriod(sliceNum, config.Period)
+	// Reset counter and compare (pwm level set to zero)
+	pwm.slice[sliceNum].ctr.ReplaceBits(0, rp.PWM_CH0_CTR_CH0_CTR_Msk, 0) // PWM_CH0_CTR_RESET
+	pwm.slice[sliceNum].cc.Set(0)                                         // PWM_CH0_CC_RESET
+
+	pwm.enable(sliceNum, start)
+}
+
+//go:inline
+func (pwm *pwms) setPhaseCorrect(sliceNum uint8, correct bool) {
+	pwm.slice[sliceNum].csr.ReplaceBits(boolToBit(correct)<<rp.PWM_CH0_CSR_PH_CORRECT_Pos, rp.PWM_CH0_CSR_PH_CORRECT_Msk, 0)
+}
+
+// Takes any of the following:
+//  rp.PWM_CH0_CSR_DIVMODE_DIV, rp.PWM_CH0_CSR_DIVMODE_FALL,
+//  rp.PWM_CH0_CSR_DIVMODE_LEVEL, rp.PWM_CH0_CSR_DIVMODE_RISE
+//go:inline
+func (pwm *pwms) setDivMode(sliceNum uint8, mode uint32) {
+	pwm.slice[sliceNum].csr.ReplaceBits(mode<<rp.PWM_CH0_CSR_DIVMODE_Pos, rp.PWM_CH0_CSR_DIVMODE_Msk, 0)
+}
+
+// setPeriod sets the pwm peripheral period (frequency). Calculates DIV_INT and sets it from following equation:
+//  cycles = (TOP+1) * (CSRPHCorrect + 1) * (DIV_INT + DIV_FRAC/16)
+// where cycles is amount of clock cycles per PWM period.
+//go:inline
+func (pwm *pwms) setPeriod(sliceNum uint8, period uint64) {
+	targetPeriod := uint32(period)
+	periodPerCycle := getPeriod()
+	top := pwm.getWrap(sliceNum)
+	phc := pwm.getPhaseCorrect(sliceNum)
```

Comment:
At the moment, phase correct PWM isn't supported so I would suggest not taking this into account.
(I believe phase correct PWM is only relevant for specific motor control applications)

Reply:
Do you mean I should leave it commented out as a TODO for when phase correct is added, if ever?